### PR TITLE
[SPARK-16674][SQL] Avoid per-record type dispatch in JDBC when reading

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

Currently, `JDBCRDD.compute` is doing type dispatch for each row to read appropriate values.
It might not have to be done like this because the schema is already kept in `JDBCRDD`.

So, appropriate converters can be created first according to the schema, and then apply them to each row.
## How was this patch tested?

Existing tests should cover this.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Could you please take a look here @cloud-fan and @yhuai ? This is happening for writing too. I will open new one for writing as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 1]
```diff
@@ -322,46 +322,134 @@ private[sql] class JDBCRDD(
...
-    case BinaryType => BinaryConversion
-    case ArrayType(et, _) => ArrayConversion(getConversions(et, metadata))
+    case BooleanType =>
+      (rs: ResultSet, pos: Int) => rs.getBoolean(pos)
+
+    case DateType =>
+      (rs: ResultSet, pos: Int) =>
+        // DateTimeUtils.fromJavaDate does not handle null value, so we need to check it.
+        val dateVal = rs.getDate(pos)
+        if (dateVal != null) {
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: `Option(dateVal).map(...).orNull`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 2]
```diff
@@ -322,46 +322,134 @@ private[sql] class JDBCRDD(
...
+          DateTimeUtils.fromJavaDate(dateVal)
+        } else {
+          null
+        }
+
+    case DecimalType.Fixed(p, s) =>
+      (rs: ResultSet, pos: Int) =>
+        val decimalVal = rs.getBigDecimal(pos)
+        if (decimalVal == null) {
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Same as above (plus you're checking equality with `null` opposite to the above -- consistency violated)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 3]
```diff
@@ -322,46 +322,134 @@ private[sql] class JDBCRDD(
...
+
+    case StringType =>
+      (rs: ResultSet, pos: Int) =>
+        // TODO(davies): use getBytes for better performance, if the encoding is UTF-8
+        UTF8String.fromString(rs.getString(pos))
+
+    case TimestampType =>
+      (rs: ResultSet, pos: Int) =>
+        val t = rs.getTimestamp(pos)
+        if (t != null) {
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: same as above

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 4]
```diff
@@ -322,46 +322,134 @@ private[sql] class JDBCRDD(
...
+          }
+
+      case dt: DecimalType =>
+        (array: Object) =>
+          array.asInstanceOf[Array[java.math.BigDecimal]].map { decimal =>
+            nullSafeConvert[java.math.BigDecimal](decimal, d => Decimal(d, dt.precision, dt.scale))
+          }
+
+      case LongType if metadata.contains("binarylong") =>
+        throw new IllegalArgumentException(s"Unsupported array element conversion.")
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Can you give more context info to the exception?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 5]
```diff
@@ -407,84 +495,8 @@ private[sql] class JDBCRDD(
         var i = 0
         while (i < conversions.length) {
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Why `while` not `foreach` or similar?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 1 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: I guess this is a critical path. I think we don't need yo introduce virtual function calls.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 5 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Using funtional transformation is generally slower due to virtual function calls. This part is executed a lot and such overhead will become really heavy. See https://github.com/databricks/scala-style-guide#traversal-and-zipwithindex

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 6]
```diff
@@ -322,46 +322,135 @@ private[sql] class JDBCRDD(
...
-  case object IntegerConversion extends JDBCConversion
-  case object LongConversion extends JDBCConversion
-  case object BinaryLongConversion extends JDBCConversion
-  case object StringConversion extends JDBCConversion
-  case object TimestampConversion extends JDBCConversion
-  case object BinaryConversion extends JDBCConversion
-  case class ArrayConversion(elementConversion: JDBCConversion) extends JDBCConversion
+  // A `JDBCConversion` is responsible for converting a value from `ResultSet`
+  // to a value in a field for `InternalRow`.
+  private type JDBCConversion = (ResultSet, Int) => Any
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: also explain what's the 2 arguments in the comment?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 7]
```diff
@@ -407,84 +496,8 @@ private[sql] class JDBCRDD(
...
-                mutableRow.update(i, null)
-              }
-            // When connecting with Oracle DB through JDBC, the precision and scale of BigDecimal
-            // object returned by ResultSet.getBigDecimal is not correctly matched to the table
-            // schema reported by ResultSetMetaData.getPrecision and ResultSetMetaData.getScale.
-            // If inserting values like 19999 into a column with NUMBER(12, 2) type, you get through
-            // a BigDecimal object with scale as 0. But the dataframe schema has correct type as
-            // DecimalType(12, 2). Thus, after saving the dataframe into parquet file and then
-            // retrieve it, you will get wrong result 199.99.
-            // So it is needed to set precision and scale for Decimal based on JDBC metadata.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: why do we remove these comments?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Oh, my mistake. I will add this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 8]
```diff
@@ -322,46 +322,135 @@ private[sql] class JDBCRDD(
...
+        if (dateVal != null) {
+          DateTimeUtils.fromJavaDate(dateVal)
+        } else {
+          null
+        }
+
+    case DecimalType.Fixed(p, s) =>
+      (rs: ResultSet, pos: Int) =>
+        val decimalVal = rs.getBigDecimal(pos)
+        if (decimalVal != null) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: use `nullSafeConvert`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 9]
```diff
@@ -322,46 +322,135 @@ private[sql] class JDBCRDD(
...
-  case object TimestampConversion extends JDBCConversion
-  case object BinaryConversion extends JDBCConversion
-  case class ArrayConversion(elementConversion: JDBCConversion) extends JDBCConversion
+  // A `JDBCConversion` is responsible for converting a value from `ResultSet`
+  // to a value in a field for `InternalRow`.
+  private type JDBCConversion = (ResultSet, Int) => Any
+
+  // This `ArrayElementConversion` is responsible for converting elements in
+  // an array from `ResultSet`.
+  private type ArrayElementConversion = (Object) => Any
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: looks like this type alias is not very useful.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 10]
```diff
@@ -407,84 +496,8 @@ private[sql] class JDBCRDD(
...
-                    throw new IllegalArgumentException("Nested arrays unsupported")
-                  case _ => array.asInstanceOf[Array[Any]]
-                }
-                mutableRow.update(i, new GenericArrayData(data))
-              } else {
-                mutableRow.update(i, null)
-              }
-          }
+          val value = conversions(i).apply(rs, pos)
+          mutableRow.update(i, value)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: previously we use `mutableRow.setXXX` for primitive types, but now we always use `update`. Is there a way we can still avoid the boxing here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 11]
```diff
@@ -322,46 +322,135 @@ private[sql] class JDBCRDD(
...
+          val data = elementConversion.apply(array)
+          new GenericArrayData(data)
+        } else {
+          null
+        }
+
     case _ => throw new IllegalArgumentException(s"Unsupported type ${dt.simpleString}")
   }
 
+  private def getArrayElementConversion(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: I'd like to inline this method, as what the previous code does

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 7 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Fxied.

**Comment by HyukjinKwon** [comment]
@cloud-fan I just addressed your comments. I added another argument in `JDBCConversion` for `MutableRow` so that we can avoid type-boxing.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 12]
```diff
@@ -322,46 +322,133 @@ private[sql] class JDBCRDD(
...
-  case object DoubleConversion extends JDBCConversion
-  case object FloatConversion extends JDBCConversion
-  case object IntegerConversion extends JDBCConversion
-  case object LongConversion extends JDBCConversion
-  case object BinaryLongConversion extends JDBCConversion
-  case object StringConversion extends JDBCConversion
-  case object TimestampConversion extends JDBCConversion
-  case object BinaryConversion extends JDBCConversion
-  case class ArrayConversion(elementConversion: JDBCConversion) extends JDBCConversion
+  // A `JDBCConversion` is responsible for converting and setting a value from `ResultSet`
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: `JDBCConversion` seems not a good name now, do you have better any ideas?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 13]
```diff
@@ -322,46 +322,133 @@ private[sql] class JDBCRDD(
...
+
+    case ArrayType(et, _) =>
+      val elementConversion = getArrayElementConversion(et, metadata)
+      (rs: ResultSet, row: MutableRow, pos: Int) =>
+        row.update(pos, nullSafeConvert(rs.getArray(pos + 1).getArray, elementConversion))
+
     case _ => throw new IllegalArgumentException(s"Unsupported type ${dt.simpleString}")
   }
 
+  private def getArrayElementConversion(dt: DataType, metadata: Metadata) = dt match {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: inline this method?

**Comment by cloud-fan** [comment]
LGTM except some style comments, thanks for working on it!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 14]
```diff
@@ -322,43 +322,130 @@ private[sql] class JDBCRDD(
...
-  case object LongConversion extends JDBCConversion
-  case object BinaryLongConversion extends JDBCConversion
-  case object StringConversion extends JDBCConversion
-  case object TimestampConversion extends JDBCConversion
-  case object BinaryConversion extends JDBCConversion
-  case class ArrayConversion(elementConversion: JDBCConversion) extends JDBCConversion
+  // A `JDBCValueSetter` is responsible for converting and setting a value from `ResultSet`
+  // into a field for `MutableRow`. The last argument `Int` means the index for the
+  // value to be set in the row and also used for the value to retrieve from `ResultSet`.
+  private type ValueSetter = (ResultSet, MutableRow, Int) => Unit
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: `JDBCValueSetter`?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Sure!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 15]
```diff
@@ -322,43 +322,130 @@ private[sql] class JDBCRDD(
...
+            s"type ${dt.simpleString} based on binary")
+
+        case ArrayType(_, _) =>
+          throw new IllegalArgumentException("Nested arrays unsupported")
+
+        case _ => (array: Object) => array.asInstanceOf[Array[Any]]
+      }
+
+      (rs: ResultSet, row: MutableRow, pos: Int) =>
+        row.update(pos, nullSafeConvert(rs.getArray(pos + 1).getArray, elementConversion))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: `new GenericArrayData`

**Comment by HyukjinKwon** [comment]
Thank you for your review @cloud-fan. Sorry for too many nits. I will try to be more careful for the next time.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 16]
```diff
@@ -322,43 +322,133 @@ private[sql] class JDBCRDD(
...
-  case object BinaryConversion extends JDBCConversion
-  case class ArrayConversion(elementConversion: JDBCConversion) extends JDBCConversion
+  // A `JDBCValueSetter` is responsible for converting and setting a value from `ResultSet`
+  // into a field for `MutableRow`. The last argument `Int` means the index for the
+  // value to be set in the row and also used for the value to retrieve from `ResultSet`.
+  private type JDBCValueSetter = (ResultSet, MutableRow, Int) => Unit
 
   /**
-   * Maps a StructType to a type tag list.
+   * Creates a StructType to setters for each type.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: update this doc?

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**